aten/src/ATen/FunctionalTensorImplBase.cpp

```diff
@@ -0,0 +1,184 @@
+  t = t.clone();
+  // This call to replace_() doesn't go through the dispatcher.
+  // It's a virtual method implemented directly on TensorImpl subclasses.
+  replace_(t.unsafeGetTensorImpl());
+  generation_ = alias_->generation();
+}
+
+bool FunctionalTensorImplBase::is_up_to_date() const {
+  if (alias_) {
+    return generation_ == alias_->generation();
+  }
+  return true;
+}
+
+void FunctionalTensorImplBase::maybe_add_update(const Tensor& updated_val) {
+  // If the mutated tensor wasn't a view, we don't need to do anything
+  if (is_view()) {
+    alias_->add_update(updated_val.clone(), view_metas_);
+    // TODO: add this back see if anything breaks?
+    //generation_ = alias_->generation(); // self is fully up to date at this point
+  }
+}
+
+void FunctionalTensorImplBase::set_view_meta(const Tensor& other, at::ViewMeta meta) {
+    auto other_impl = dynamic_cast<at::FunctionalTensorImplBase*>(other.unsafeGetTensorImpl());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(other_impl != nullptr);
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(alias_ == nullptr);
+    if (other_impl->alias_ != nullptr) {
+        // if the other tensor is already a view, copy its ViewMeta vector and push the current one.
+        std::vector<at::ViewMeta> metas = other_impl->view_metas_;  // copy
+        metas.push_back(meta);
+        //alias_ = other_impl->alias_; // refcount bump
+        alias_ = other_impl->alias_; // refcount bump
+        view_metas_ = metas;
+    } else {
+        std::shared_ptr<Alias> alias = std::make_shared<Alias>(const_cast<Tensor&>(other));
+        at::ViewMeta noop_view_info(at::ViewMeta::Type::noOp, other.sizes().vec(), other.sizes().vec());
+        // The original tensor wasn't a view - turn it into a (no-op) view.
+        other_impl->alias_ = alias;
+        other_impl->view_metas_ = std::vector<at::ViewMeta>{noop_view_info};
+        // Turn the new tensor into a view too
+        alias_ = alias;
+        view_metas_ = std::vector<at::ViewMeta>{meta};
+    }
+}
+
+bool FunctionalTensorImplBase::is_view() const {
+    return view_metas_.size() != 0;
+}
+
+const char* FunctionalTensorImplBase::tensorimpl_type_name() const {
+    return "FunctionalTensorImplBase";
+}
+
+namespace functionalization {
+namespace impl {
+
+void maybe_add_update(Tensor& self) {
+  auto functional_base_impl = dynamic_cast<at::FunctionalTensorImplBase*>(self.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(functional_base_impl != nullptr);
+  functional_base_impl->maybe_add_update(self);
+}
+
+void set_view_meta(const at::Tensor& out, const at::Tensor& t, ViewMeta meta) {
+  auto out_impl = dynamic_cast<at::FunctionalTensorImplBase*>(out.unsafeGetTensorImpl());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(out_impl != nullptr);
+  out_impl->set_view_meta(t, meta);
+}
+
+void set_view_meta(const c10::List<at::Tensor>& outs, const at::Tensor& t, ViewMeta meta) {
+  for (auto& out: outs.vec()) {
```

Comment:
Why we converting the c10::List to a vec? Just iterate over it directly...

Reply:
I think I just tried writing `for (auto& out: outs) {...}` and the compiler complained - probably should have marked with a todo to come back to it.

I think this should work:
```
for (const auto i : c10::irange(outs.size())) {* use outs[i] *}
```